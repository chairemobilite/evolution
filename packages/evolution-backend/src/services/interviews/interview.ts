/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import _set from 'lodash.set';
import _unset from 'lodash.unset';
import moment from 'moment';
import { UserAttributes } from 'chaire-lib-backend/lib/services/auth/user';
import serverValidate, { ServerValidation } from '../validations/serverValidation';
import serverUpdateField, { ServerFieldUpdateCallback } from './serverFieldUpdate';
import config from 'chaire-lib-backend/lib/config/server.config';
import interviewsDbQueries from '../../models/interviews.db.queries';
import { UserInterviewAttributes, InterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import projectConfig from '../../config/projectConfig';

export const addRolesToInterview = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>,
    user: UserAttributes
) => {
    // Add the userRoles in the interview object
    const permissions = user.permissions;
    interview.userRoles = permissions ? Object.keys(permissions).filter((perm) => permissions[perm] === true) : [];
};

export const setInterviewFields = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    interview: InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>,
    options: {
        valuesByPath: { [key: string]: unknown };
        unsetPaths?: string[];
    }
) => {
    // Set the interview's values by path received from client
    for (const path in options.valuesByPath) {
        const value = options.valuesByPath[path];
        _set(interview, path, value);
    }
    // Unset any interview path defined by client
    if (Array.isArray(options.unsetPaths)) {
        for (let i = 0, count = options.unsetPaths.length; i < count; i++) {
            _unset(interview, options.unsetPaths[i]);
        }
    }
};

/**
 * Update an interview, given the paths to update. If database logs are
 * configured, a new log will be saved for this update.
 *
 * TODO: Rename to save, as it conveys more the fact that it is saved in the db than update
 *
 * @param {InterviewAttributes} interview The base interview to update, as
 * fetched in the database
 * @param {{valuesByPath: { [key: string]: unknown }; unsetPaths?: string[];
 * serverValidations?: ServerValidation; fieldsToUpdate?: (keyof
 * InterviewAttributes)[]; logData?: { [key: string]: unknown };}} options The
 * `valuesByPath` is a key-value map of values to update in the interview. The
 * key is the dot-separated hierarchical path to the field to update. For
 * example, to update a field in response, it could be `{
 * 'responses.mySection.myField': 'myValue' }`. The `unsetPaths` array is an
 * array of dot-separated paths whose value to unset. The 'serverValidations'
 * are the validations to do on the data. If any new field to set does not
 * validate, the value will be set in the responses object, but a corresponding
 * value will be set to false in the validations object. The 'fieldsToUpdate'
 * makes sure only those fields are updated in the interview. Since there is no
 * control on the paths of the valuesByPath keys, it avoids updating fields in
 * code path where they should not be updated. The 'logData' object contains
 * data that should be added to the log object of the current update, if log is
 * enabled.
 **/
export const updateInterview = async <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    interview: InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>,
    options: {
        logDatabaseUpdates?: boolean;
        valuesByPath: { [key: string]: unknown };
        unsetPaths?: string[];
        serverValidations?: ServerValidation;
        fieldsToUpdate?: (keyof InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>)[];
        logData?: { [key: string]: unknown };
    }
): Promise<{
    interviewId: string | undefined;
    serverValidations:
        | true
        | {
              [key: string]: {
                  [key: string]: string;
              };
          };
    serverValuesByPath: { [key: string]: unknown };
}> => {
    const fieldsToUpdate = options.fieldsToUpdate || ['responses', 'validations'];
    const logData = options.logData || {};
    const serverValidations = await serverValidate(
        options.serverValidations,
        options.valuesByPath,
        options.unsetPaths || []
    );
    // Update values by path with caller provided values
    setInterviewFields(interview, { valuesByPath: options.valuesByPath, unsetPaths: options.unsetPaths });
    const serverValuesByPath = await serverUpdateField(
        interview,
        projectConfig.serverUpdateCallbacks,
        options.valuesByPath,
        options.unsetPaths
    );
    // Update values with server updated values
    if (Object.keys(serverValuesByPath).length > 0) {
        setInterviewFields(interview, { valuesByPath: serverValuesByPath, unsetPaths: options.unsetPaths });
    }
    // If server validation failed, add the invalid fields to validation
    if (serverValidations !== true) {
        for (const path in serverValidations) {
            _set(interview.validations, `${path}`, false);
        }
    }
    interview.responses = interview.responses || {};
    interview.validations = interview.validations || {};
    if (!interview.logs || !Array.isArray(interview.logs)) {
        interview.logs = [];
        //console.log(interview.logs);
    }

    const databaseUpdateJson: Partial<InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>> =
        {};
    fieldsToUpdate.forEach((field) => {
        // FIXME: For some reason, the following line gives type error, not sure why, so casting to any and disabling the warning.
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (databaseUpdateJson as any)[field] = interview[field];
    });

    // logs if configured:
    // FIXME Type evolution's configs
    if (options.logDatabaseUpdates !== false && (config as any).logDatabaseUpdates) {
        const timestamp = interview.responses._updatedAt || moment().unix();
        if (options.unsetPaths) {
            interview.logs.push({
                ...logData,
                timestamp,
                valuesByPath: options.valuesByPath,
                unsetPaths: options.unsetPaths
            });
        } else {
            interview.logs.push({
                ...logData,
                timestamp,
                valuesByPath: options.valuesByPath
            });
        }
        databaseUpdateJson.logs = interview.logs;
    }

    const retInterview = await interviewsDbQueries.update(interview.uuid, databaseUpdateJson);
    return { interviewId: retInterview.uuid, serverValidations, serverValuesByPath };
};
