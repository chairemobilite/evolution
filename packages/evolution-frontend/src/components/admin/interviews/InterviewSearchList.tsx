/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';

import Interview from './InterviewSearchResult';
import { InterviewContext } from '../../../contexts/InterviewContext';
import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import InterviewsCreateLink from './InterviewsCreateLink';

type InterviewSearchListProps = {
    autoCreateIfNoData: boolean;
    queryData: URLSearchParams;
};

const InterviewSearchList: React.FunctionComponent<InterviewSearchListProps & WithTranslation> = (
    props: InterviewSearchListProps & WithTranslation
) => {
    const [data, setData] = React.useState<{ [key: string]: any }[]>([]);
    const [loading, setLoading] = React.useState(false);
    const { state, dispatch } = React.useContext(InterviewContext);
    const fetchIdRef = React.useRef(0);

    // Function to fetch data from the server, with paging and filtering
    const fetchData = React.useCallback(
        async (accessCode: string) => {
            // Give this fetch an ID
            const fetchId = ++fetchIdRef.current;

            // Set the loading state
            setLoading(true);

            // Make a query string from the filters

            try {
                const response = await fetch(`/api/interviews/interviewByCode?accessCode=${accessCode}`);
                if (fetchId !== fetchIdRef.current) {
                    // There was another query since, ignore
                    return;
                }
                if (response.status === 200) {
                    const jsonData = await response.json();
                    if (jsonData.interviews) {
                        setData(jsonData.interviews);
                        if (
                            !_isBlank(accessCode) &&
                            jsonData.interviews.length === 0 &&
                            props.autoCreateIfNoData &&
                            state.status !== 'creating'
                        ) {
                            // automatically create new interview with this access code
                            dispatch({
                                type: 'createNew',
                                username: `telephone_${(Math.ceil(Math.random() * 8999) + 1000).toString()}`,
                                queryData: props.queryData
                            });
                        }
                    }
                } else {
                    console.error('Invalid response code from server: ', response.status);
                }
            } catch (error) {
                console.error(`Error fetching user data from server: ${error}`);
                setData([]);
            } finally {
                if (fetchId === fetchIdRef.current) {
                    setLoading(false);
                }
            }
        },
        [state.status]
    );

    React.useEffect(() => {
        fetchData(state.responses.accessCode);
    }, [state.responses]);

    // TODO This is very specific to the interviewer use case, generalize it and use a function to generate username, maybe as prop
    return (
        <ul className="interview">
            {data.map((interview) => (
                <Interview key={interview.uuid} interview={interview} />
            ))}
            <InterviewsCreateLink queryData={props.queryData} />
        </ul>
    );
};

export default withTranslation(['admin', 'main'])(InterviewSearchList);
