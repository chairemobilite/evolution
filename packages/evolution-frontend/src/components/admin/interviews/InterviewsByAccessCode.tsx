/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { useTranslation } from 'react-i18next';
import Loadable from 'react-loadable';
import Loader from 'react-spinners/HashLoader';

import { InterviewContext } from '../../../contexts/InterviewContext';
import InputString from 'chaire-lib-frontend/lib/components/input/InputString';
import { Link, useLocation, useParams } from 'react-router';
import { _booleish } from 'chaire-lib-common/lib/utils/LodashExtensions';

const loader = function Loading() {
    return <Loader size={30} color={'#aaaaaa'} loading={true} />;
};

const InterviewsComponent = Loadable({
    // TODO: move this to componentDidMount like in Monitoring.tsx
    loader: () => import('./InterviewSearchList'),
    loading: loader
});

const InterviewsByAccessCode: React.FC = () => {
    const location = useLocation();
    const params = useParams();
    const { t } = useTranslation('admin');

    const urlSearch = new URLSearchParams(location.search);
    const [currentCode, setCurrentCode] = React.useState(params.accessCode);
    const [createNewIfNoData] = React.useState(_booleish(urlSearch.get('autoCreate')) === true);
    const { state, dispatch } = React.useContext(InterviewContext);

    React.useEffect(() => {
        if (params.accessCode !== state.response.accessCode) {
            dispatch({ type: 'update_response', response: { accessCode: params.accessCode } });
        }
    }, [params.accessCode, state.response]);

    return (
        <div className="admin">
            <div className="survey-section__content apptr__form-container">
                <InputString
                    id="accessCodeSearchInput"
                    value={currentCode}
                    onValueUpdated={(newValue) => setCurrentCode(newValue.value)}
                ></InputString>
                <Link to={`/interviews/byCode/${currentCode}`}>
                    <button type="button" className={'survey-section__button button blue small'}>
                        {t('admin:interviewSearch:SearchByCode')}
                    </button>
                </Link>
            </div>
            <InterviewsComponent autoCreateIfNoData={createNewIfNoData} queryData={urlSearch} />
        </div>
    );
};

export default InterviewsByAccessCode;
