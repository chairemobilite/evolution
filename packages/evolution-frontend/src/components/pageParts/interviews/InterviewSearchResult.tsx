/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { WithTranslation, withTranslation } from 'react-i18next';

interface InterviewProps extends WithTranslation {
    interview: { [key: string]: any };
    key: string;
}

const InterviewSearchResult: React.FunctionComponent<InterviewProps> = (props: InterviewProps) => {
    const interviewUrl = `/survey/edit/${props.interview.uuid}/`;
    return (
        <li
            title={props.interview.uuid}
            key={props.key}
            className={`
                ${props.interview.isValid === true && props.interview.isCompleted === true ? '_dark-green _strong' : ''}
                ${
        props.interview.isValidated === true &&
                    props.interview.isValid === true &&
                    props.interview.isCompleted === true
            ? '_green _strong _active-background'
            : ''
        }
                ${props.interview.isValid === true && props.interview.isCompleted === false ? '_orange _strong' : ''}
                ${props.interview.isValid === false ? '_dark-red _strong' : ''}
                `}
        >
            {props.t('admin:interviewSearch:InterviewUser')}
            {': '}
            {props.interview.email ||
                props.interview.username ||
                (props.interview.facebook
                    ? props.t('admin:interviewSearch:FromFacebook')
                    : props.interview.google
                        ? props.t('admin:interviewSearch:FromGoogle')
                        : props.t('admin:interviewSearch:UnknownUser'))}{' '}
            •{' '}
            {props.interview.isCompleted === true
                ? props.t('admin:CompletedFemSingular')
                : props.interview.isCompleted === false
                    ? props.t('admin:NotCompletedFemSingular')
                    : props.t('admin:UnknownCompletionFemSingular')}{' '}
            •{' '}
            {props.interview.isValid === true
                ? props.t('admin:Valid')
                : props.interview.isValid === false
                    ? props.t('admin:Invalid')
                    : props.t('admin:UnknownValidity')}{' '}
            • {props.interview.home.address || props.t('admin:interviewSearch:UnknownAddress')},{' '}
            {props.interview.home.city || ''} •{' '}
            <a
                href={interviewUrl}
                id={`interviewButtonList_${props.interview.uuid}`}
                data-uuid={props.interview.uuid}
                target="_blank"
                rel="noreferrer"
            >
                {props.t('admin:interviewSearch:Open')}
            </a>
        </li>
    );
};

export default withTranslation(['admin', 'main'])(InterviewSearchResult);
