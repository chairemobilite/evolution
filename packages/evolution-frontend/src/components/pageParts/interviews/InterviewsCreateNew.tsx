/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import { createBrowserHistory } from 'history';
import Loader from 'react-spinners/HashLoader';
import { InterviewContext } from '../../../contexts/InterviewContext';

const InterviewsCreateNew: React.FunctionComponent = () => {
    const [inProgress, setInProgress] = React.useState(false);
    const { state, dispatch } = React.useContext(InterviewContext);

    // Function to fetch data from the server, with paging and filtering
    const createUser = React.useCallback(async () => {
        if (inProgress || state.status !== 'creating') {
            return;
        }

        // Set in progress state
        setInProgress(true);

        // Make a query string from the filters
        try {
            const response = await fetch('/api/interviews/createNew', {
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json'
                },
                credentials: 'include',
                method: 'POST',
                body: JSON.stringify({
                    createUser: state.username,
                    responses: state.responses
                })
            });
            const history = createBrowserHistory({ forceRefresh: true });
            if (response.status === 200) {
                const jsonData = await response.json();
                if (jsonData.interviewUuid) {
                    history.push(`/survey/edit/${jsonData.interviewUuid}/`);
                    dispatch({ type: 'success', interviewUuid: jsonData.interviewUuid });
                    return;
                }
                console.log('Server did not return a proper interview id', jsonData);
                // Goto proper error page
            }
            // TODO Go to proper status code page
            console.log('Bad response code from server', response.status);
        } catch (error) {
            // TODO Go to proper error page
            console.error(`Error fetching user data from server: ${error}`);
        }

        // Cancel in progress state, only if the page was not redirected
        setInProgress(false);
    }, []);

    React.useEffect(() => {
        createUser();
    }, []);

    return (
        <div className="admin">
            {inProgress && <Loader size={30} color={'#aaaaaa'} loading={true} />}
            {!inProgress && 'error'}
        </div>
    );
};

/** Component used to create new interviews by users other than the main user */
export default InterviewsCreateNew;
