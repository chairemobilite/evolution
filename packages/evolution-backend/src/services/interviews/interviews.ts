/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { auditInterview, getChangesAfterCleaningInterview } from 'evolution-common/lib/services/interviews/interview';
import { updateInterview, setInterviewFields } from './interview';
import { mapResponsesToValidatedData } from './interviewUtils';

import { validateAccessCode } from '../accessCode';
import { validate as validateUuid } from 'uuid';
import interviewsDbQueries, { InterviewSearchAttributes, OperatorSigns } from '../../models/interviews.db.queries';
import {
    InterviewAttributes,
    InterviewListAttributes,
    UserInterviewAttributes
} from 'evolution-common/lib/services/interviews/interview';

const getFiltersForDb = (
    filter: { is_valid?: 'valid' | 'invalid' | 'notInvalid' | 'notValidated' | 'all' } & {
        [key: string]: string | { value: string | boolean | number | null; op?: keyof OperatorSigns };
    }
): {
    [key: string]: { value: string | boolean | number | null; op?: keyof OperatorSigns };
} => {
    const { is_valid, ...filters } = filter;

    const actualFilters: {
        [key: string]: { value: string | boolean | number | null; op?: keyof OperatorSigns };
    } = {};

    // Add the desired validity query
    switch (is_valid) {
    case 'valid':
        actualFilters.is_valid = { value: true, op: 'eq' };
        break;
    case 'invalid':
        actualFilters.is_valid = { value: false, op: 'eq' };
        break;
    case 'notInvalid':
        actualFilters.is_valid = { value: false, op: 'not' };
        break;
    case 'notValidated':
        actualFilters.is_valid = { value: null, op: 'eq' };
        break;
    default:
        // No filter required
        break;
    }

    Object.keys(filters).forEach((key) => {
        const filter = filters[key];
        if (typeof filter === 'string') {
            actualFilters[key] = { value: filter };
        } else {
            actualFilters[key] = filter;
        }
    });

    return actualFilters;
};

export default class Interviews {
    // TODO The actual location of the access code may depend on the questionnaire. It should not be hard-coded here, but passed by the actual application.
    static findByAccessCode = async (accessCode: string): Promise<InterviewSearchAttributes[]> => {
        if (!validateAccessCode(accessCode)) {
            return [];
        }
        return await interviewsDbQueries.findByResponse({ accessCode });
    };

    static getInterviewByUuid = async <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
        interviewId: string
    ): Promise<InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> | undefined> => {
        if (!validateUuid(interviewId)) {
            return undefined;
        }
        return await interviewsDbQueries.getInterviewByUuid(interviewId);
    };

    static getUserInterview = async <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
        userId: number
    ): Promise<UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> | undefined> => {
        return await interviewsDbQueries.getUserInterview(userId);
    };

    static createInterviewForUser = async <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
        participantId: number,
        initialResponses: { [key: string]: any },
        returning: string | string[] = 'uuid'
    ): Promise<Partial<InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>>> => {
        // TODO Make sure there is no active interview for this user already?

        // Create the interview for this user
        const interview = await interviewsDbQueries.create(
            { participant_id: participantId, responses: initialResponses, is_active: true, validations: {}, logs: [] },
            returning
        );
        if (!interview.uuid || Object.keys(initialResponses).length === 0) {
            return interview as InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
        }
        // update interview with initial responses so that server updates are run on the initial responses
        const userInterview = await Interviews.getInterviewByUuid(interview.uuid);
        if (userInterview === undefined) {
            throw 'Interview just created was not found!';
        }
        const valuesByPath = {};
        Object.keys(initialResponses).forEach((key) => {
            valuesByPath[`responses.${key}`] = initialResponses[key];
        });
        await updateInterview(userInterview, {
            valuesByPath,
            fieldsToUpdate: ['responses']
        });
        return interview as InterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    };

    // TODO Add filters fields as required
    static getAllMatching = async <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
        params: {
            filter?: { is_valid?: 'valid' | 'invalid' | 'notInvalid' | 'notValidated' | 'all' } & {
                [key: string]: string | { value: string | boolean | number | null; op?: keyof OperatorSigns };
            };
            pageIndex?: number;
            pageSize?: number;
            updatedAt?: number;
            sort?: (string | { field: string; order: 'asc' | 'desc' })[];
        } = {}
    ): Promise<{
        interviews: InterviewListAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>[];
        totalCount: number;
    }> => {
        const pageIndex = params.pageIndex || 0;
        const pageSize = params.pageSize || -1;
        const updatedAt = params.updatedAt || 0;

        const actualFilters = getFiltersForDb(params.filter || {});
        // Add the updated at query if the time is larger than 0
        if (updatedAt > 0) {
            actualFilters.updated_at = { value: updatedAt, op: 'gte' };
        }

        return await interviewsDbQueries.getList({ filters: actualFilters, pageIndex, pageSize, sort: params.sort });
    };

    static getValidationErrors = async (
        params: {
            filter?: { is_valid?: 'valid' | 'invalid' | 'notInvalid' | 'notValidated' | 'all' } & {
                [key: string]: string | { value: string | boolean | number | null; op?: keyof OperatorSigns };
            };
        } = {}
    ): Promise<{ errors: { key: string; cnt: string }[] }> => {
        const actualFilters = getFiltersForDb(params.filter || {});

        return await interviewsDbQueries.getValidationErrors({ filters: actualFilters });
    };

    static auditInterviews(validations, surveyProjectHelper, parsers): Promise<void> {
        const queryStream = interviewsDbQueries.getInterviewsStream({ filters: {} });
        let i = 0;
        return new Promise((resolve, reject) => {
            queryStream
                .on('error', (error) => {
                    console.error('queryStream failed', error);
                    reject(error);
                })
                .on('data', (row) => {
                    queryStream.pause();
                    // Pausing the connnection is useful if your processing involves I/O
                    const interview = row;
                    const changesAfterCleaningInterview = mapResponsesToValidatedData(
                        getChangesAfterCleaningInterview(
                            interview.validated_data,
                            interview.responses,
                            interview,
                            parsers,
                            surveyProjectHelper
                        ),
                        []
                    );
                    setInterviewFields(interview, changesAfterCleaningInterview);
                    const audits = auditInterview(
                        interview.validated_data,
                        interview.responses,
                        interview,
                        validations,
                        surveyProjectHelper
                    );
                    changesAfterCleaningInterview.valuesByPath.audits = audits;
                    updateInterview(interview, {
                        fieldsToUpdate: ['audits'],
                        logDatabaseUpdates: false,
                        valuesByPath: changesAfterCleaningInterview.valuesByPath
                    })
                        .then(() => {
                            queryStream.resume();
                        })
                        .catch((error) => {
                            console.error(error);
                            queryStream.end();
                            reject(error);
                        });
                    console.log(i + 1, interview.uuid);
                    i++;
                })
                .on('end', () => {
                    console.log('all interviews audited successfully');
                    resolve();
                });
        });
    }

    static resetInterviews(confirm: string): Promise<void> {
        if (confirm !== 'I WANT TO DELETE ALL VALIDATION WORK') {
            return new Promise((_resolve, reject) => {
                reject('The confirm string should be \'I WANT TO DELETE ALL VALIDATION WORK\'');
            });
        }
        const queryStream = interviewsDbQueries.getInterviewsStream({ filters: {} });
        let i = 0;
        return new Promise((resolve, reject) => {
            queryStream
                .on('error', (error) => {
                    console.error('queryStream failed', error);
                    reject(error);
                })
                .on('data', (row) => {
                    queryStream.pause();
                    // Pausing the connnection is useful if your processing involves I/O
                    const interview = row;
                    updateInterview(interview, {
                        fieldsToUpdate: ['validated_data', 'is_completed', 'is_validated', 'is_valid'],
                        logDatabaseUpdates: false,
                        valuesByPath: {
                            validated_data: interview.responses,
                            is_completed: null,
                            is_validated: null,
                            is_valid: null
                        }
                    })
                        .then(() => {
                            queryStream.resume();
                        })
                        .catch((error) => {
                            console.error(error);
                            queryStream.end();
                            reject(error);
                        });
                    console.log(i + 1, interview.uuid);
                    i++;
                })
                .on('end', () => {
                    console.log('all interviews reset successfully');
                    resolve();
                });
        });
    }

    static statEditingUsers = async (
        params: { permissions?: string[] } = {}
    ): Promise<{ email: string; count: number }[]> => {
        return await interviewsDbQueries.statEditingUsers(params);
    };
}
