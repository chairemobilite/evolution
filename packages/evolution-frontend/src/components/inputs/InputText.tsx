/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';

import { _isBlank } from 'chaire-lib-common/lib/utils/LodashExtensions';
import { InputTextType } from 'evolution-common/lib/services/widgets';
import { parseString } from 'evolution-common/lib/utils/helpers';
import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import { CliUser } from 'chaire-lib-common/lib/services/user/userType';

export interface InputTextProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> {
    id: string;
    onValueChange?: (e: any) => void;
    value?: string;
    interview: UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    // TODO There's also a path in widgetConfig, but this one comes from the props of the question. See if it's always the same and use the one from widgetConfig if necessary
    path: string;
    user: CliUser;
    inputRef?: React.LegacyRef<HTMLTextAreaElement>;
    widgetConfig: InputTextType<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
}

export const InputText = <CustomSurvey, CustomHousehold, CustomHome, CustomPerson>(
    props: InputTextProps<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>
) => (
        <textarea
            autoComplete="off"
            rows={props.widgetConfig.rows || 5}
            className={'apptr__form-input input-text'}
            name={props.widgetConfig.shortname}
            id={props.id}
            defaultValue={
                _isBlank(props.value)
                    ? parseString(props.widgetConfig.defaultValue, props.interview, props.path, props.user)
                    : props.value
            }
            onBlur={props.onValueChange}
            ref={props.inputRef}
            maxLength={props.widgetConfig.maxLength || 2000}
        />
    );

export default InputText;
