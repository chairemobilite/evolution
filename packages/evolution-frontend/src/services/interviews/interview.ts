import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import { WidgetConfig, LangData } from '../widgets';

export type WidgetStatus<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> = {
    path: string;
    customPath?: string;
    isVisible: boolean;
    modalIsOpen: boolean;
    isDisabled: boolean;
    isCollapsed: boolean;
    isEmpty: boolean;
    isCustomEmpty: boolean;
    isValid: boolean;
    isResponded: boolean;
    isCustomResponded: boolean;
    errorMessage?: LangData<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    groupedObjectId?: string;
    value: unknown;
    customValue?: unknown;
    /** Key that should be changed if the widget is updated outside the application workflow, eg. from the server */
    currentUpdateKey: number;
};

/* FIXME This type represents what was in the legacy evolution, with fields that
 evolve throughout the application, they represent state of the local interview
 more than interview attributes. This all should be moved to some class, with
 methods to follow the state of the current interview. */
export type FrontendInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> = {
    previousWidgets?: {
        [widgetShortname: string]: WidgetStatus<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
    };
    previousGroups?: {
        [groupName: string]: {
            [groupId: string]: {
                [widgetShortname: string]: WidgetStatus<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
            };
        };
    };
    // These are widgets statuses for the current section, if they are not grouped
    widgets: { [widgetShortname: string]: WidgetStatus<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> };
    // These are the widget status for the groups in the current section
    groups: {
        [groupName: string]: {
            [groupId: string]: {
                [widgetShortname: string]: WidgetStatus<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
            };
        };
    };
    // Contains the paths in the responses of the visible widgets... FIXME Rename?
    visibleWidgets: string[];
    allWidgetsValid: boolean;
    // Name of the currently loaded section
    sectionLoaded?: string;
};

export type UserFrontendInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> =
    FrontendInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> &
        UserInterviewAttributes<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;

export type SurveySectionGroup = {
    widgets: string[];
};

// TODO Properly type this
export type SurveySection = {
    widgets: string[];
    groups?: {
        [groupName: string]: SurveySectionGroup;
    };
    [key: string]: unknown;
};
export type SurveySections = { [sectionName: string]: SurveySection };
export type SurveyWidgets<CustomSurvey, CustomHousehold, CustomHome, CustomPerson> = {
    [widgetName: string]: WidgetConfig<CustomSurvey, CustomHousehold, CustomHome, CustomPerson>;
};
