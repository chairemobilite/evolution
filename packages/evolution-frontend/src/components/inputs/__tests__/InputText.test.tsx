/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import { UserInterviewAttributes } from 'evolution-common/lib/services/interviews/interview';
import React from 'react';
import TestRenderer from 'react-test-renderer';

import { InputText } from '../InputText';

const userAttributes = {
    id: 1,
    username: 'foo',
    preferences: {  },
    serializedPermissions: [],
    isAuthorized: () => true,
    is_admin: false,
    pages: [],
    showUserInfo: true
}

const interviewAttributes: UserInterviewAttributes = {
    id: 1,
    uuid: 'arbitrary uuid',
    user_id: 1,
    is_completed: false,
    responses: {
        section1: {
            q1: 'abc',
            q2: 3
        },
        section2: {
            q1: 'test'
        }
    },
    validations: {
        section1: {
            q1: true,
            q2: false
        },
        section2: {
            q1: true
        }
    },
    is_valid: true
};

test('Should correctly render InputText with all parameters', () =>{

    const widgetConfig = {
        type: 'question' as const,
        twoColumns: true,
        path: 'test.foo',
        inputType: 'text' as const,
        datatype: 'text' as const,
        maxLength: 1500,
        defaultValue: 'test',
        shortname: 'testName',
        rows: 10,
        containsHtml: true,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    }

    const wrapper = TestRenderer.create(
        <InputText
            id={'test'}
            onValueChange={() => { /* nothing to do */}}
            widgetConfig={widgetConfig}
            value='value'
            inputRef={React.createRef()}
            size='medium'
            interview={interviewAttributes}
            path={'path'}
            user={userAttributes}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('Should correctly render InputText with base parameters', () =>{

    const widgetConfig = {
        type: 'question' as const,
        path: 'test.foo',
        inputType: 'text' as const,
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    }
    const wrapper = TestRenderer.create(
        <InputText
            id={'test'}
            widgetConfig={widgetConfig}
            value='value'
            interview={interviewAttributes}
            path={'path'}
            user={userAttributes}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('Should correctly render InputText with defaultValue', () =>{

    const widgetConfig = {
        type: 'question' as const,
        path: 'test.foo',
        inputType: 'text' as const,
        defaultValue: 'This is the text that should be present',
        label: {
            fr: `Texte en français`,
            en: `English text`
        }
    }
    const wrapper = TestRenderer.create(
        <InputText
            id={'test'}
            widgetConfig={widgetConfig}
            value=''
            interview={interviewAttributes}
            path={'path'}
            user={userAttributes}
        />
    );
    expect(wrapper).toMatchSnapshot();
});
