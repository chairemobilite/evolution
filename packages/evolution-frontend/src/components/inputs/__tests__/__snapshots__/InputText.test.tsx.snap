// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should correctly render InputText with all parameters 1`] = `
<textarea
  autoComplete="off"
  className="apptr__form-input input-text"
  defaultValue="value"
  id="test"
  maxLength={1500}
  name="testName"
  onBlur={[Function]}
  rows={10}
/>
`;

exports[`Should correctly render InputText with base parameters 1`] = `
<textarea
  autoComplete="off"
  className="apptr__form-input input-text"
  defaultValue="value"
  id="test"
  maxLength={2000}
  rows={5}
/>
`;

exports[`Should correctly render InputText with defaultValue 1`] = `
<textarea
  autoComplete="off"
  className="apptr__form-input input-text"
  defaultValue="This is the text that should be present"
  id="test"
  maxLength={2000}
  rows={5}
/>
`;
